experimenter/experimenter/experiments/migrations/0266_nimbusexperiment_results_pairwise_data.py

```diff
@@ -0,0 +1,181 @@
+# Generated by Django 3.2.23 on 2024-01-24 21:10
+
+from copy import deepcopy
+from django.db import migrations, models
+
+
+def update_results_data_schema(apps, schema_editor):
+    NimbusExperiment = apps.get_model("experiments", "NimbusExperiment")
+
+    for experiment in NimbusExperiment.objects.all():
+        branches = [b.slug for b in experiment.branches.all()]
+        reference_branch = experiment.reference_branch
+        data = experiment.results_data
+        if data is not None and "v3" not in data and "v2" in data:
+            data["v3"] = data["v2"]
+            """
+            data for v3 looks like this (with changes vs v2 noted):
+            [window]: {
+                [analysis_basis]: {
+                    [segment]: {
+                        [branch]: {
+                            "branch_data": {
+                                [metrics_group]: {
+                                    [metric_name]: {
+                                        "absolute": {
+                                            "all": [...],
+                                            "first": {...}
+                                        },
+                                        "difference": {
+                                            // "all" and "first" moved from here into [comparison_branch]
+                                            [comparison_branch]: { // this level is new
+                                                "all": [...],
+                                                "first": {...},
+                                            },
+                                        },
+                                        "relative_uplift": {
+                                            // "all" and "first" moved from here into [comparison_branch]
+                                            [comparison_branch]: { // this level is new
+                                                "all": [...],
+                                                "first": {...},
+                                            },
+                                        },
+                                        "significance": {
+                                            // optional daily/weekly/overall moved from here into [comparison_branch]
+                                            [comparison_branch]: { // this level is new
+                                                "weekly": {},
+                                                "overall: {}
+                                            }
+                                        },
+                                        "percent": 0.0
+            """
+
+            comparison_default = {"first": {}, "all": []}
+            significance_default = {"weekly": {}, "overall": []}
+
+            for key, value in data["v3"].items():
```

Comment:
Please pull some of this out into helper functions. The nesting makes it unreadable.

Reply:
I too hate this function. I started trying to do this recursively, but the dynamic nature of all of the nested keys made it extremely difficult to follow, and I was unsure if it would even work. @jaredlockhart's original suggestion was to not spend a lot of time on it because a migration only runs once. In light of the errors merging this the past few times I can take another pass at reorganizing it, but I'm not sure the best way to go about it.

Reply:
Pulling out the contents of this loop into a function would be a good start.

Reply:
Well the flip side of this is, we tested this locally against the prod data in its current form and validated the output.  Applying a bunch of structural refactoring now could potentially introduce new bugs.  This code only runs once, we don't need to worry about the maintainability of it.  Yes the structure is unreadable and difficult to reason about, but I'd rather focus on finding ways to improve this pattern in general for future code/migrations than spend time trying to apply them here.